{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read and translate numbers to integers in text\n",
    "\n",
    "### Cases considered:\n",
    "* number words only: three million two hundred three\n",
    "* number words with \"and\": three million two hundred and three\n",
    "* two separate numbers separated by word other than \"and\": two hundred chickens and three hens\n",
    "* Mixed numbers and words: 3.4 million\n",
    "* Numbers with commas: 2,514,200\n",
    "\n",
    "### TODO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "from pdfminer.pdfinterp import PDFResourceManager, PDFPageInterpreter\n",
    "from pdfminer.converter import TextConverter\n",
    "from pdfminer.layout import LAParams\n",
    "from pdfminer.pdfpage import PDFPage\n",
    "import os\n",
    "import sys, getopt\n",
    "\n",
    "#converts pdf, returns its text content as a string\n",
    "def convert(fname, pages=None):\n",
    "    if not pages:\n",
    "        pagenums = set()\n",
    "    else:\n",
    "        pagenums = set(pages)\n",
    "\n",
    "    output = StringIO()\n",
    "    manager = PDFResourceManager()\n",
    "    converter = TextConverter(manager, output, laparams=LAParams())\n",
    "    interpreter = PDFPageInterpreter(manager, converter)\n",
    "\n",
    "    infile = open(fname, 'rb')\n",
    "    for page in PDFPage.get_pages(infile, pagenums):\n",
    "        interpreter.process_page(page)\n",
    "    infile.close()\n",
    "    converter.close()\n",
    "    text = output.getvalue()\n",
    "    output.close\n",
    "    return text \n",
    "\n",
    "#converts all pdfs in directory pdfDir, saves all resulting txt files to txtdir\n",
    "def convertMultiple(pdfDir, txtDir):\n",
    "    if pdfDir == \"\": \n",
    "        pdfDir = os.getcwd() + \"\\\\\" #if no pdfDir passed in \n",
    "    for pdf in os.listdir(pdfDir): #iterate through pdfs in pdf directory\n",
    "        fileExtension = pdf.split(\".\")[-1]\n",
    "        if fileExtension == \"pdf\":\n",
    "            pdfFilename = pdfDir + pdf \n",
    "            text = convert(pdfFilename) #get string of text content of pdf\n",
    "#             textFilename = txtDir + pdf + \".txt\"\n",
    "#             textFile = open(textFilename, \"w\") #make text file\n",
    "#             textFile.write(text) #write text to text file\n",
    "    return text\n",
    "\n",
    "# set paths accordingly:\n",
    "pdfDir = \"/Users/peterschnatz/1HC/pdf_reading_test/pdf_files/\"\n",
    "txtDir = \"\"\n",
    "text = convertMultiple(pdfDir, txtDir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Designing heterotropically activated allosteric conformational switches using supercharging  Peter J', ' Schnatza, Joseph M', ' Brisendineb, Craig C', ' Lainga, Bernard H', ' Eversona, Cooper A', ' Frenchb, Paul M', ' Molinaroa, and Ronald L', ' Kodera,c,d,e,f,1\\ue840  aDepartment of Physics, The City College of New York, New York, NY 10031; bDepartment of Biochemistry, The City College of New York, New York, NY 10031; cGraduate Program of Physics, The Graduate Center, The City University of New York, New York, NY 10016; dGraduate Program of Biology, The Graduate Center, The City University of New York, New York, NY 10016; eGraduate Program of Biochemistry, The Graduate Center, The City University of New York, New York, NY 10016; and fGraduate Program of Chemistry, The Graduate Center, The City University of New York, New York NY 10016  Edited by William F', ' DeGrado, University of California, San Francisco, CA, and approved January 27, 2020 (received for review September 19, 2019)  Heterotropic allosteric activation of protein function, in which binding of one ligand thermodynamically activates the binding of another, different ligand or substrate, is a fundamental control mechanism in metabolism and as such has been a long-aspired capability in protein design', ' Here we show that greatly increasing the magnitude of a protein’s net charge using surface supercharging transforms that protein into an allosteric ligand- and counterion-gated conformational molecular switch', ' To demonstrate this we first modi- fied the designed helical bundle hemoprotein H4, creating a highly charged protein which both unfolds reversibly at low ionic strength and undergoes the ligand-induced folding transition commonly ob- served in signal transduction by intrinsically disordered proteins in biology', ' As a result of the high surface-charge density, ligand binding to this protein is allosterically activated up to 1,300-fold by low con- centrations of divalent cations and the polyamine spermine', ' To extend this process further using a natural protein, we similarly modified Escherichia coli cytochrome b562 and the resulting protein behaves in a like manner', ' These simple model systems not only establish a set of general engineering principles which can be used to convert natural and designed soluble proteins into allosteric molecular switches useful in biodesign, sensing, and synthetic biology, the behavior we have demonstrated––functional activation of supercharged intrinsically dis- ordered proteins by low concentrations of multivalent ions––may be a control mechanism utilized by Nature which has yet to be appreciated', ' intrinsically disordered proteins | supercharging | allostery | ligand-induced folding | protein design Intrinsically disordered proteins (IDPs) in nature commonly  function by making a structural change from an unfolded, disordered state to a folded, functional state in response to binding another protein or small molecule or as a result of phosphorylation or dephosphorylation (1)', ' This process, in which the binding of another molecule to an IDP induces a phase change to an active structure, is termed “ligand-induced folding” and is a form of heterotropic allosteric activation now known to be prevalent in natural IDPs (2)', '  The hallmarks of IDP structure are a high net charge per residue (3), a relatively low core protein-folding energy (4), and low sequence diversity (5)', ' IDPs are found in almost every functional signal transduction pathway in biology, including cell cycling (6), circadian rhythms (6), posttranscriptional regulation (7), and protein degradation (8)', ' This universality points to the utility of the phase change in regulation: as form equals function in much of molecular biology, having the ability to turn form itself, and therefore function, on and off is a simple and attrac- tive method to control individual nodes in complex interaction networks', ' The ability to design switchable IDPs, both by modi- fying extant natural proteins and by fully de novo IDP design, is critical to the success of future complex network creation efforts in synthetic biology', '  Here we demonstrate that the modification of a protein’s sur- face to greatly increase the magnitude of its net charge imparts the  Y G O L O I B  L A N O I T A T U P M O C  D N A  S C I S Y H P O I B  ligand-induced folding behavior found in natural IDPs', ' Further, the high negative charge density imparts heterotropic allosteric ligand binding behavior in response to divalent cation and polyamino compound effector molecules––specific ion interactions that have not yet been observed in high net-charge natural IDPs', ' Heterotropic allostery is a long-discussed goal in protein design (9)', ' This work thus presents a roadmap for the refashioning of natural and designed globular proteins into allosteric molecular switches which will prove useful in the design of control mechanisms in natural and designed metabolic networks and in the fashioning of improved biosensors', '  Results H4 Protein Redesign', ' The starting point is the designed single-chain diheme tetrahelical bundle protein H4, which is a single-chain variant of the homodimeric protein HP-7 (10, 11)', ' Apo-H4 has a surplus of negatively charged residues––16 more acidic than basic residues (Fig', ' 1A)', ' Because of this, the apoprotein partially unfolds at low ionic strength (SI Appendix, Fig', ' S1)', ' We set out to amplify this property by increasing the net charge', ' The software package ProtCAD (12) was used to create a structural model for holo-H4, and every fully surface-exposed sidechain was changed to glutamic acid, creating an H4 variant termed H4(-52) with 52 excess neg- atively charged sidechains', ' H4 also contains low helical-propensity aspartate and asparagine sidechains that were originally added to increase chemical shift dispersion (13), and these were reverted to  Significance  We present a general method to imbue proteins with hetero- tropic allostery by using surface protein supercharging to impart intrinsic disorder to the protein fold, allowing for functional ac- tivation by low concentrations of multivalent ions, including Mg(II), Ca(II), and spermine', ' We have designed two different proteins which exhibit this property and present a protocol which anyone can employ to impart this property to their pro- tein of interest', ' This method promises to enable future synthetic biology projects which utilize natural fluxes of these ions to di- rectly actuate function at significantly faster rates than that of genetic activation', ' Such multivalent ionic activation of disor- dered proteins may be a mechanism utilized by Nature which has yet to be appreciated', '  Author contributions: P', 'J', 'S', ', J', 'M', 'B', ', B', 'H', 'E', ', C', 'A', 'F', ', and R', 'L', 'K', ' designed research; P', 'J', 'S', ', J', 'M', 'B', ', C', 'C', 'L', ', B', 'H', 'E', ', C', 'A', 'F', ', and P', 'M', 'M', ' performed research; P', 'J', 'S', ', J', 'M', 'B', ', B', 'H', 'E', ', C', 'A', 'F', ', P', 'M', 'M', ', and R', 'L', 'K', ' analyzed data; and P', 'J', 'S', ', J', 'M', 'B', ', and R', 'L', 'K', ' wrote the paper', '  The authors declare no competing interest', '  This article is a PNAS Direct Submission', '  Published under the PNAS license', ' 1To whom correspondence may be addressed', ' Email: rkoder@ccny', 'cuny', 'edu', '  This article contains supporting information online at https://www', 'pnas', 'org/lookup/suppl/ doi:10', '1073/pnas', '1916046117/-/DCSupplemental', '  www', 'pnas', 'org/cgi/doi/10', '1073/pnas', '1916046117  PNAS Latest Articles  | 1 of 7     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0cFig', ' 1', ' Supercharging of H4 and E', ' coli cytochrome b562 and model thermodynamic scheme', ' Alpha carbon representation of (A) H4, (B) H4(-28), (C) wild-type cytochrome b562, and (D) b562(-22)', ' Corresponding amino acid sequences are displayed below each structure with helical regions underlined', ' Acidic residues are colored red and basic residues are colored blue', ' (E) Equilibrium thermodynamic cycle for stepwise folding and binding, and ligand-induced folding for H4(-28)', '  their high-propensity analogs glutamate and glutamine', ' This pro- tein did not fold even at high sodium chloride concentrations (SI Appendix, Fig', ' S2)', ' Fifty-two acidic residues comprise 40% of the protein, a net-charge-per-residue higher than the 35% cutoff value identified by Pappu and coworkers as the random coil limit (3)', ' We then created the protein H4(-28) by changing every lysine residue in H4 to glutamine and hand-selecting a subset of surface-exposed residues to become acidic such that 21% of the protein sidechains are negatively charged, there are no basic residues except for the four heme ligand histidines, and the charged surface sidechains are maximally distant from each other (Fig', ' 1B)', ' This net-charge-per- residue value lies in the region predicted by Pappu and coworkers to exhibit ligand-induced folding behavior, and unlike H4(-52), high concentrations of sodium chloride fold the protein as detected by circular dichroism', '  The NaCl Dependence of the Folding Energy of Apo-H4(-28)', ' In 20 mM borate buffer, pH 9', '3, apo-H4(-28) has a circular dichroism spectrum indicative of a predominantly random coil structure at NaCl activi- ties less than 200 mM', ' Fig', ' 2A depicts the NaCl-induced folding transition of H4(-28) as detected by circular dichroism: helices form as the NaCl activity is increased, with the protein becoming pre- dominantly helical at concentrations over 1 M', ' 15N-HSQC spectra of  the apoprotein at low NaCl are consistent with an extended random coil conformation while the apoprotein at high NaCl has NMR spectra consistent with that of a molten globule, the same structural state that the apoprotein form of H4’s progenitor protein, HP7, forms (Fig', ' 3) (13)', ' Likewise, 1H-1H-NOESY (nuclear Overhauser effect spectra) display a few sharp cross-peaks in the low-salt sample, as expected for a random coil protein, while there is a large field of broad, low cross-peaks in the high-salt sample as expected for a molten globular protein undergoing intermediate exchange', ' A sim- ilar conclusion can be drawn from 1-anilino-naphthalene-8-sulfonate (ANS) binding to H4(-28) (14): at low NaCl ANS does not bind to the protein but at NaCl activities high enough to promote helix formation ANS binding is apparent (SI Appendix, Fig', ' S3A)', '  Fig', ' 2B depicts urea denaturation experiments conducted at higher NaCl concentrations between 750 and 2,000 mM', ' To- gether with the circular dichroism data these demonstrate that the unfolded-to-molten globule transition energy of apo-H4(-28) depends linearly on NaCl concentration over a range of 8 kcal/mol as the electrostatic screening increases (blue and red triangles, Fig', ' 2E)', ' The continuous decrease of the folding energy well into the regime where nearly the entire ensemble is condensed indicates that screening effects influence the stability of the molten globular structure with equal magnitude as when the structure energetically  2 of 7 | www', 'pnas', 'org/cgi/doi/10', '1073/pnas', '1916046117  Schnatz et al', '     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0cY G O L O I B  L A N O I T A T U P M O C  D N A  S C I S Y H P O I B  Fig', ' 2', ' Folding and binding equilibrium thermodynamics for H4(-28)', ' (A) CD spectra for H4(-28) at 20 mM borate pH 9', '3, and NaCl concentrations increasing from 50 to 2,500 mM', ' (A, Inset) Molar residue ellipticity at 222 nm', ' (B) Urea denaturation curves of H4(-28) at 20 mM borate pH 9', '3, and NaCl activities ranging from 0', '75 to 2 M', ' (C) CD spectra for H4(-28) titrated with IHP at 20 mM borate pH 9', '3 and 300 mM NaCl', ' (B, Inset) Binding curve showing molar residue ellipticity at 222 nm', ' (D) Dissociation constants of H4(-28) and IHP as a function of NaCl activity', ' (E) Folding and binding energies of H4(-28) and IHP as a function of NaCl activity at 20 mM borate pH 9', '3', ' All error bars, average ± SD, n = 3', '  favors the unfolded state', ' This suggests that only the solvent-exposed residues are influenced by ionic strength and the core residues, which are not well-solvated when apo-H4(-28) is folded, are unaf- fected by changes in NaCl concentration', ' Importantly, at low NaCl concentrations the folding energy is positive, offering the opportu- nity for folding to be triggered by the coupled stabilizing energy of cofactor binding', '  Ligand-Induced Folding of H4(-28)', ' Thermodynamic measurements of heme binding were not successful at high NaCl concentrations because heme forms irreversible aggregates on a minutes time- scale (SI Appendix, Fig', ' S4)', ' Instead the heme analog Fe(III)- isohematoporphyrin (IHP) (Fig', ' 2C) was used, which was found to be soluble for hours at NaCl concentrations up to 2 M', ' In 20 mM borate, 300 mM NaCl, pH 9', '3 buffer––a NaCl concentration  at which the apoprotein is predominantly unfolded––two mol- ecules of IHP bind independently to apo-H4(-28) with a dissociation constant of 30 ± 5 μM (Fig', ' 2C)', ' No cooperativity was detected at any NaCl concentration, a result identical to that observed for the progenitor protein HP-7 (13)', ' Circular dichroism spectra of the protein during the titration dem- onstrate helical formation concomitant with heme binding (Fig', ' 2C, Inset)', ' NMR analysis of the holoprotein under the same conditions demonstrates the degree of chemical shift resolution consistent with a natively folded tetrahelical bun- dle (Fig', ' 3), and ANS fluorescence is likewise reduced (SI Appendix, Fig', ' S3B)', ' Thus under these solution conditions H4(-28) exhibits the ligand-induced folding behavior, going from an extended random coil structure to a nativelike folded structure, seen in natural IDPs', '  Schnatz et al', '  PNAS Latest Articles  | 3 of 7     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0creveal the intrinsic interaction energy, ΔGdock, to be −9', '5 ± 0', '1 kcal/mol', ' When plotted together (Fig', ' 2E), it is obvious that the values of ΔGbind,app and ΔGF1 as a function of NaCl concentration are par- allel within error', ' The average distance between the two lines is −7', '1 ± 0', '5 kcal/mol', ' ΔGF1 is the apoprotein unfolded-to-molten globule transition energy as determined using circular dichroism in Fig', ' 2 A and B', ' This energy can be used in conjunction with Kd,app and Eq', ' 3 to calculate ΔGF2:  KF2 =  KdockðKF1 + 1Þ \\x00 Kd,app − Kdock KF1  \\x01', '  [3]  The derived values of ΔGF2 also vary linearly with NaCl with a slope identical within error to that of both ΔGF1and ΔGbind,app and an average distance of 1', '9 ± 0', '3 kcal/mol higher than ΔGF1', ' Thus we have shown that for supercharged H4(-28), the effective binding constant can be controlled with a dynamic range of more than two orders of magnitude using the NaCl concentration, sug- gesting that for a given NaCl concentration the effective binding constant could likewise be controlle d over a similar range by adjusting the surface charge density', '  Supercharging the Natural Cytochrome b562', ' H4 is an artificially designed protein', ' In order to demonstrate the broad scope of the supercharging effect, we applied the same design algorithm to the natural cytochrome b562 (Fig', ' 1 C and D)', ' Wild-type b562 differs from H4 in that it has a unique structure in both the apo- and holo states (16)', ' There are a number of conserved positive residues near the heme binding site (17) and we retained those while changing a number of surface-exposed residues outside the heme binding site to glutamic acid', ' The final sequence, b562(-22), has a net charge-per-residue of −0', '21, identical to that of H4(-28)', ' As Fig', ' 4A depicts, b562(-22) exhibits NaCl-dependent folding behavior similar to that of H4(-28) with the difference that it takes more than a twofold greater concentration of NaCl for sufficient charge screening to shift the equilibrium to the folded state', '  The evolved heme binding site of b562(-22) is specific for heme—IHP cannot be used as a soluble analog', ' The use of heme in ligand binding experiments limits experiments to NaCl concen- trations less than 500 mM', ' Fig', ' 4B depicts an optically detected heme binding experiment at 500 mM NaCl, a NaCl concentration at which the apoprotein is primarily unfolded', ' As Fig', ' 4C dem- onstrates, binding of heme under the same conditions is accom- panied by a great increase in the helical content of the protein', ' Fig', ' 4D depicts the linear NaCl dependence of both the folding and binding energies of b562(-22)', ' Although heme solubility limits the range of experimentally accessible binding energies, we ob- serve that the energies again are parallel lines with a separa- tion of −12', '6 ± 0', '2 kcal/mol, slightly less than the affinity reported for wild-type b562 of −11', '0 kcal/mol (18)', ' As wild-type apo-b562 has a disordered region that is ordered upon heme binding (16), the 1', '6-kcal/mol difference might be the energetic cost of ordering this region', '  The Exaggerated Stabilizing Effect of Multivalent Cations', ' The high density of negative residues on the protein surface engenders an exaggerated response to multivalent cations', ' As Fig', ' 5A depicts, Ca2+ and Mg2+ ions induce H4(-28) folding at concentrations two orders of magnitude lower than Na+ and Zn2+ and the polyamine spermine induce folding at H4(-28) concentrations even two orders of magnitude lower than that', ' Fig', ' 5B compares the ionic strength dependence of the folding energy of H4(-28) in both NaCl and CaCl2 solutions', ' Unlike with NaCl, the response to Ca2+ is biphasic––at low ionic strengths, the protein is stabi- lized by greater than 26-fold more by Ca2+ solutions than Na+', ' At ionic strengths greater than 80 mM, stabilization by both is  Fig', ' 3', ' NMR spectra for three phases of H4(-28): 700 MHz 1H-15N-HSQC spectra of apo-H4(-28) at 20 mM monosodium phosphate pH 7', '0 with (A) 50 mM NaCl and (B) 750 M NaCl; and (C) ferric heme-bound holo-H4(-28) with 100 mM NaCl', ' 1H-1H-NOESY at 700 MHz of apo-H4(-28) in 20 mM bo- rate pH 9', '3 with (D), 50 mM NaCl and (E), 1 M NaCl; and (F), ferric heme- bound holo-H4(-28) with 300 mM NaCl', '  Coupled Binding and Folding in H4(-28)', ' The apparent binding en- ergy of the ligand under a particular solution condition is the intrinsic interaction energy between the ligand and the folded protein (ΔGdock) increased by the energetic cost required to shift the protein population completely to the folded state (15)', ' As H4(-28) has three folding states,  ΔGbind,app = ΔGdock + ΔGF1 + ΔGF2,  [1]  where ΔGF1 and ΔGF2 are the apoprotein unfolded-to-molten globule and molten globule-to-folded transition energies (Fig', ' 1E)', ' From these coupled equilibria it follows that (SI Appendix)  \\x00\\x00  \\x01  Kd,app = Kd,dock  K−1 F1  p K−1 F2  + K−1 F2  \\x01 + 1  ,  [2]  which predicts how the apparent binding energy varies as a function of the folding equilibria of the apoprotein', ' The intrinsic dissocia- tion constant, Kd,dock, is a fixed parameter for the cofactor-binding site pair', ' In supercharged proteins such as H4(-28), the addition of negatively charged residues to the surface of the folded structure makes ΔGF1 and ΔGF2 tunable parameters sensitive to the ionic strength of the environment', ' As the NaCl concentration increases, ΔGF2 approaches zero, and ΔGbind,app = ΔGdock', ' These coupled folding and binding equilibria are illustrated in Fig', ' 2D by the increase in Kd,app with decreasing NaCl', ' As Eq', ' 2 predicts and Fig', ' 2E demonstrates, increasing NaCl activities exponentially decrease the apparent dissociation constant up to 1 M NaCl by decreasing the energetic cost of folding, and the apparent binding energy decreases linearly with Debye length over that range', ' Binding experiments performed at higher NaCl  4 of 7 | www', 'pnas', 'org/cgi/doi/10', '1073/pnas', '1916046117  Schnatz et al', '     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0cbinding becomes too tight to accurately measure', ' While the narrow dynamic range possible with heme precludes the kind of large-scale exploration of binding thermodynamics we have performed for H4(-28), we do observe similar allosteric behavior by cytochrome b562(-22) (SI Appendix, Fig', ' S6)', ' Discussion By supercharging the protein surface, we have instilled a form of heterotropic allosteric modulation via the remodeling of con- formational heterogeneity seen in natural intrinsically disordered proteins (2) on both a designed and natural hemoprotein', ' In both cases we have made it so that the net charge of the protein is −0', '21 per residue and demonstrated that we can modulate the folding energy, and thus the effective binding constant, by the addition of different cations, suggesting that for a given cation concentration the effective binding constant could likewise be controlled over a similar range by adjusting the surface charge density', ' Allosteric activation was at least as large as 4', '4 kcal/mol, corresponding to a greater than 1,300-fold increase in affinity in the case of Ca2+ ion, sufficient for such supercharged proteins to act as molecular switches actuated by polyvalent cations', ' These proteins further serve as useful model systems, or maquettes, for natural IDP structure and function', ' It seems likely that some  Y G O L O I B  L A N O I T A T U P M O C  D N A  S C I S Y H P O I B  Fig', ' 5', ' Folding and binding equilibrium thermodynamics with polyvalent cations', ' (A) Folding energy as a function of titrant concentration for H4(-28) at 20 mM Mops pH 7', '0', ' (B) Folding energy as a function of ionic strength for H4(-28) at 20 mM Mops pH 7', '0 and NaCl (black) or CaCl2 (red)', ' SEs of linear fit parameters calculated by the matrix inversion method', ' (C) Folding energy as a function of Debye length for H4(-28) at 20 mM Mops pH 7', '0', ' (D) Folding energy as a function of titrant concentration for b562(-22) at 20 mM Mops pH 7', '0', ' (E) CD spectra for H4(-28) at the divalent cation concentrations used to show allosteric activation in F and the spectra of the fully folded protein at high concentrations of divalent cation', ' (F) Ultraviolet-visible spectra for H4(-28) titrated with IHP at 20 mM Mops pH 7', '0, 50 mM NaCl, and low con- centrations of divalent cations to activate allosteric binding', ' (F, Inset) Zoomed-in view of binding curves with divalent cations', '  Fig', ' 4', ' Folding and binding equilibrium thermodynamics for b562(-22)', ' (A) CD spectra of b562(-22) in 20 mM Mops pH 7', '0 and NaCl activities increasing from 15 to 3000 mM', ' (A, Inset) Molar residue ellipticity at 222 nm', ' (B) Optical spectra of b562(-22) titrated with heme at 20 mM Mops pH 7', '0 and 500 mM NaCl', ' (B, Inset) Binding curve showing absorption at 418 nm', ' (C) CD spectra of b562(-22) titrated with IHP at 20 mM Mops pH 7', '0 and 500 mM NaCl', ' (Inset) Binding curve showing molar residue ellipticity at 222 nm', ' (D) Folding and binding energy of b562(-22) and heme as a function of NaCl activity at 20 mM Mops pH 7', '0', ' SEs of linear fit parameters calculated by the matrix inversion method', '  equivalent within error', ' The spacing between the parallel por- tions of the response is 3', '7 ± 0', '3 kcal/mol', '  This biphasic behavior suggests a model in which the first phase represents stabilization via Ca2+ binding to clusters of an- ionic sidechains while the second phase represents further stabili- zation by charge screening', ' That the second phase is a function of charge screening is supported by the fact that it has the same slope within error as that of charge-screening stabilization by NaCl', ' Cation binding in the first phase is strongly supported by the Debye length analysis of the energetic response: when the folding energy at a series of low cation concentrations is plotted against the Debye length of each solution (Fig', ' 5C), each multivalent ion has a linear response that significantly differs from that of NaCl, and indeed different from each other', ' Stabilization by Debye screening alone would result in plots that are superimposable', ' b562(-22) has a similarly exaggerated response to multivalent cations (Fig', ' 5D), demonstrating that this property can also be imparted to natural proteins', '  Allosteric Activation of Cofactor Binding by Multivalent Cations', ' Multivalent cation binding to H4(-28) allosterically activates IHP binding by lowering the thermodynamically coupled folding energy barrier', ' Fig', ' 5E depicts the circular dichroism spectra of the protein both in the presence of moderate concentrations of multivalent cations, insufficient to effect detectable folding of apo-H4(-28), and at relatively high concentrations of those same cations, those suf- ficient to shift the equilibrium fully to the molten globular state', ' Fig', ' 5F depicts the significant increase in cofactor binding affinity, as much as 36-fold, produced by the introduction of small con- centrations of these effectors', ' Zn2+ was shown to competitively inhibit cofactor binding, suggesting that it at least in part binds in the cofactor binding pocket (SI Appendix, Fig', ' S5)', '  At high concentrations these effector molecules activate cofactor binding as much as 1,300-fold, activation energies large enough to serve as a molecule switch (Table 1)', ' The actual maximal activation energies are even higher, as at higher cation concentrations cofactor  Schnatz et al', '  PNAS Latest Articles  | 5 of 7     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0cTable 1', ' Binding enhancement by allosteric effector cations  Ion concentration  50 mM NaCl 50 mM NaCl 1', '5 mM MgCl2 50 mM NaCl 1', '0 mM CaCl2 50 mM NaCl 60 μM spermine 50 mM NaCl 15 mM MgCl2 50 mM NaCl 12', '5 mM CaCl2 50 mM NaCl 6 mM spermine  Kd, μm 8 ± 1  0', '52 ± 0', '03  0', '22 ± 0', '02  0', '27 ± 0', '03  Binding enhancement  15 ± 2  36 ± 6  30 ± 5  0', '013 ± 0', '004  600 ± 200  0', '006 ± 0', '001  1,300 ± 300  0', '008 ± 0', '002  1,000 ± 300  ΔGapp  bind, kcal/mol −7', '0 ± 0', '1 −8', '68 ± 0', '06  −9', '20 ± 0', '09  −9', '1 ± 0', '1  −10', '9 ± 0', '3  −11', '4 ± 0', '2  −11', '2 ± 0', '3  ΔΔG, kcal/mol  −1', '7 ± 0', '1  −2', '2 ± 0', '1  −2', '1 ± 0', '1  −3', '9 ± 0', '3  −4', '4 ± 0', '2  −4', '2 ± 0', '3  highly charged natural IDPs are likewise activated by polyvalent cation fluxes', ' Furthermore, our proteins resemble halophilic proteins in their high fraction of surface acidic groups (19), and it is possible that the behavior we observe here may also be exhibited by the highly charged natural proteins seen in halophiles', '  This process opens a pathway to the creation of proteins, en- zymes, and metabolic pathways with folding and assembly triggered by the addition, creation, or transport of polyvalent cations', ' Ca2+ influx, for example, has been shown to be constitutively driven by the mechanosensitive channels of Escherichia coli when they undergo mechanical stress induced by hydrogels (20), promising the ability to activate specifically designed proteins and pathways in bacterial films with spatial resolution', ' Photocaged metal complexes offer the ability to allosterically activate such proteins with both spatial and temporal control (21)', ' Likewise, the high sensitivity to polyamines promises the ability to activate such pathways by in- ducing any of several polyamine biosynthetic pathways (22)', ' Such multivalent ionic activation promises to enable synthetic biology experiments that use ion fluxes to directly actuate function at sig- nificantly faster rates than that of genetic activation', '  Furthermore, engineered biosensors may substantially benefit from having a structural phase transition accompanying the binding of a target analyte', ' The changes in bulk material prop- erties that result from the phase transition, such as the index of refraction in a monolayer of receptor proteins, offer the poten- tial for large signal-to-noise gains over current biosensor tech- nology, and of recreating naturally occurring molecular signaling mechanisms in an artificial context (23)', '  Materials and Methods Materials', ' Hemin was purchased from Fluka', ' IHP was purchased from Frontier Scientific', ' Stock solutions of cofactors were freshly prepared at 1 mg/mL in dimethyl sulfoxide and used within 4 h', ' Proteins were expressed and purified as we have published before (24), with the addition of a final high- performance liquid chromatography purification step using a C18 reverse-phase column (Western Analytical)', '  Spectroscopy', ' Optical spectra were recorded with a Hewlett-Packard 8452A diode array spectrophotometer using a 1-cm path length glass cuvette', ' Circular dichroism (CD) spectra were recorded with an Aviv Biomedical CD Spectrometer model 400 in a 0', '2-cm light path quartz cell, using a band- width of 1–2 nm, an averaging time of 2 s, and averaged over two scans', ' Spectra were recorded between 200 and 250 nm, with background correc- tion', ' Fluorescence spectra were recorded using an Olis DM45 fluorescence spectrometer using a 2', '5 × 10-mm quartz cell', '  NaCl-Driven Folding', ' H4(-28) experiments with NaCl were performed at 20 mM borate pH 9', '3', ' H4(-28) experiments with multivalent ions were performed at 20 mM 3-(N-morpholino)propanesulfonic acid (Mops) pH 7', '0', ' b562(-22) ex- periments were performed at 20 mM Mops pH 7', '0', ' Apoprotein was pre- pared at 2', '5–10 μM in buffer with incremental NaCl concentrations noted', '  The 222-nm band was used to determine the alpha helical content at each NaCl concentration and the linear extrapolation method was used to cal- culate the folding energy as a function of ionic strength', '  Chemical Denaturation', ' Solutions of 2', '5 μM apo-H4(-28) and apo-b562(-22) were prepared in buffers noted with urea concentrations ranging from 0 to 8 M', ' CD spectra showed no significant change when the samples were equilibrated for 24 h, hence equilibration time was not strictly monitored and was normally on the order of minutes', ' Spectra were recorded between 210 and 250 nm, with background correction', ' The folding energy was determined by fitting the urea melt curves to the equation  ½θ\\x8a  =  obs  ½θ\\x8a  Ue−ðΔG0−m½D\\x8aÞ=RT  +½θ\\x8a 1 + e−ðΔG0−m½D\\x8aÞ=RT  N  ', '  [4]  Here, ΔG0 is the folding energy, m is an empirical constant determining how the stability changes with denaturant concentration, ½D\\x8a is the denaturant concentration, R is the universal gas constant, and T is the temperature', ' Also, ½θ\\x8a = ξU½D\\x8a + ξ0,U determine how the signal of the na- tive and unfolded states change with denaturant concentration, where ξ0,i are the y intercepts of the native and denatured baselines and ξi are the slopes of the baselines', '  = ξN½D\\x8a + ξ0,N and ½θ\\x8a  N  U  Cofactor Binding–Optical', ' For experiments with varying NaCl concentrations, apo-H4(-28) was prepared at 0', '5 μM in 20 mM borate pH 9', '3 and noted NaCl concentrations', ' For experiments with polyvalent cations, apo-H4(-28) was prepared at 0', '175–0', '5 μM in 20 mM Mops pH 7', '0', ' Apo-b562 (-22) was pre- pared at 1 μM in 20 mM Mops pH 7', '0 and noted NaCl concentration', ' Suc- cessive 0', '1–5 equivalent additions of cofactor were followed by an equilibration period ranging from 1', '5 to 5 min', ' Optical spectra were col- lected after each equilibration period and the cofactor binding was moni- tored using the bound cofactor Soret peak', ' Values for Kd were obtained from plots of the Soret band absorbance vs', ' the concentration of cofactor added and fit with least-squares regression analysis using the tight-binding equation: A = A0 + eunb½cof\\x8a 2 4Kd +½cof\\x8a +½prot\\x8a − + ebnd½prot\\x8a  ﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃﬃ 3 ðKd +½cof\\x8a +½prot\\x8aÞ2 − 4½cof\\x8a½prot\\x8a 5,  q  2½prot\\x8a  [5]  where A0 is the initial absorbance of the Soret band, eunb is the molar ab- sorption coefficient of unbound cofactor at the Soret band peak absorbance, eunb is the additional Soret band absorbance of bound cofactor, [cof] is the cofactor concentration, [prot] is the protein concentration, and Kd is the dis- sociation constant for oxidized cofactor at a specified NaCl concentration', '  Cofactor Binding–CD', ' At low NaCl concentrations, cofactor binding was monitored via CD spectrometry', ' Apoprotein samples were 2', '5 μ M in 20 mM boric acid pH 9', '3 for H4(-28) and 20 mM Mops pH 7', '0 for b562(-22) with the specified NaCl concentration', ' Successive 0', '25–50 equivalent additions of IHP were followed by a 5-min equilibration period', ' CD spectra were collected after each equilibration period and the cofactor binding was monitored by  6 of 7 | www', 'pnas', 'org/cgi/doi/10', '1073/pnas', '1916046117  Schnatz et al', '     0 2 0 2    ,  5 2     I  I          y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  \\x0cthe ellipticity at 222 nm, ½θ\\x8a the ½θ\\x8a  222 vs', ' the concentration of IHP least-squares fit with Eq', ' 5', '  222', ' Values for Kd were determined from plots of  NMR Spectroscopy: 1H-15N-HSQC', ' Phase-sensitive gradient-enhanced 2D 1H-15N heteronuclear single quantum coherence (HSQC) spectra (25) were collected at 25 °C on a Bruker AVANCE III HD 700-MHz NMR spectrometer at 700', '2 MHz for protons and 70', '9 MHz for 15N with sweep widths of 11,160 Hz for 1H and 1,987 Hz for 15N utilizing globally optimized alternating-phase (GARP) decoupling of 15N during 1H acquisition', ' Chemical shifts were referenced to water at 4', '77 ppm for 1H and external 15NH4Cl in 1 M HCl at 20 °C at 24', '93 ppm for 15N', ' A total of 256 t1 increments of 2,048 complex data points were collected with eight scans per t1 increment and the relaxation delay was 1 s', ' The total acquisition time was ∼40 min', ' Two-dimensional (2D) spectra were processed using Bruker’s TopSpin 3', '2 software package', '  NMR Spectroscopy: 1H-1H-NOESY', ' Two-dimensional NOESY were collected at 25 °C on a Bruker AVANCE III HD 700-MHz NMR spectrometer at 700', '2 MHz  for protons with sweep widths of 11,160 Hz for t1 and 12,276 Hz for t2', ' A total of 400 t1 increments of 2,048 complex data points were collected with 16 scans per t1 increment and the relaxation delay was 1', '5 s', ' The mixing time was 100 ms and the total acquisition time was ∼3 h 10 min', ' Two-dimensional spectra were processed using Bruker’s TopSpin 3', '2 software package', '  Data Availability', ' All data are included in the main text and SI Appendix', '  ACKNOWLEDGMENTS', ' The authors would like to thank Doug Barrick (Department of Biophyiscs, Johns Hopkins University) for valuable discus- sions regarding the thermodynamics of these systems', ' R', 'L', 'K', ' gratefully acknowledges support via National Institutes of Health Grant GM111932', ' B', 'H', 'E', ' gratefully acknowledges support from the Center for Exploitation of Nanostructures in Sensor and Energy Systems under NSF Cooperative Agreement Award 0833180', ' R', 'L', 'K', ' is a member of the New York Structural Biology Center (NYSBC)', ' NMR data collected at NYSBC was made possible by a grant from NYSTAR', '  1', ' H', ' J', ' Dyson, P', ' E', ' Wright, Intrinsically unstructured proteins and their functions', ' Nat', '  Rev', ' Mol', ' Cell Biol', ' 6, 197–208 (2005)', '  2', ' H', ' N', ' Motlagh, J', ' O', ' Wrabl, J', ' Li, V', ' J', ' Hilser, The ensemble nature of allostery', ' Nature  508, 331–339 (2014)', '  3', ' A', ' H', ' Mao, S', ' L', ' Crick, A', ' Vitalis, C', ' L', ' Chicoine, R', ' V', ' Pappu, Net charge per residue modulates conformational ensembles of intrinsically disordered proteins', ' Proc', ' Natl', ' Acad', ' Sci', ' U', 'S', 'A', ' 107, 8183–8188 (2010)', '  4', ' E', ' B', ' Gibbs, S', ' A', ' Showalter, Quantitative biophysical characterization of intrinsically  disordered proteins', ' Biochemistry 54, 1314–1326 (2015)', '  5', ' P', ' Romero et al', ', Sequence complexity of disordered protein', ' Proteins 42, 38–48  (2001)', '  6', ' C', ' A', ' Galea, Y', ' Wang, S', ' G', ' Sivakolundu, R', ' W', ' Kriwacki, Regulation of cell division by Intrinsic flexibility, modularity, and signaling  intrinsically unstructured proteins: conduits', ' Biochemistry 47, 7598–7609 (2008)', '  7', ' A', ' Bah, J', ' D', ' Forman-Kay, Modulation of intrinsically disordered protein function by  post-translational modifications', ' J', ' Biol', ' Chem', ' 291, 6696–6705 (2016)', '  15', ' K', ' L', ' Chan et al', ', Characterization of the Zn(II) binding properties of the human Wilms’ tumor suppressor protein C-terminal zinc finger peptide', ' Inorg', ' Chem', ' 53, 6309–6320 (2014)', '  16', ' Y', ' Feng, S', ' G', ' Sligar, A', ' J', ' Wand, Solution structure of apocytochrome b562', ' Nat', '  Struct', ' Biol', ' 1, 30–35 (1994)', '  17', ' C', ' Negron, C', ' Fufezan, R', ' L', ' Koder, Helical templates for porphyrin binding in de-  signed proteins', ' Proteins 74, 400–416 (2009)', '  18', ' C', ' R', ' Robinson, Y', ' Liu, J', ' A', ' Thomson, J', ' M', ' Sturtevant, S', ' G', ' Sligar, Energetics of heme binding to native and denatured states of cytochrome b562', ' Biochemistry 36, 16141– 16146 (1997)', '  19', ' G', ' Ortega, T', ' Diercks, O', ' Millet, Halophilic protein adaptation results from synergistic residue-ion interactions in the folded and unfolded states', ' Chem', ' Biol', ' 22, 1597–1607 (2015)', '  20', ' G', ' N', ' Bruni, R', ' A', ' Weekley, B', ' J', ' T', ' Dodd, J', ' M', ' Kralj, Voltage-gated calcium flux me- diates Escherichia coli mechanosensation', ' Proc', ' Natl', ' Acad', ' Sci', ' U', 'S', 'A', ' 114, 9445–9450 (2017)', '  8', ' R', ' van der Lee et al', ', Intrinsically disordered segments affect protein half-life in the  21', ' H', ' W', ' Mbatia, S', ' C', ' Burdette, Photochemical tools for studying metal ion signaling  cell and during evolution', ' Cell Rep', ' 8, 1832–1844 (2014)', '  and homeostasis', ' Biochemistry 51, 7212–7224 (2012)', '  9', ' V', ' Nanda, R', ' L', ' Koder, Designing artificial enzymes by intuition and computation', ' Nat', '  22', ' A', ' J', ' Michael, Polyamines in eukaryotes, bacteria, and archaea', ' J', ' Biol', ' Chem', ' 291,  Chem', ' 2, 15–24 (2010)', '  14896–14903 (2016)', '  10', ' A', ' C', ' Mutter et al', ', Rational design of a zinc phthalocyanine binding protein', ' J', ' Struct', '  Biol', ' 185, 178–185 (2014)', '  11', ' L', ' Zhang et al', ', Manipulating cofactor binding thermodynamics in an artificial oxygen  transport protein', ' Biochemistry 50, 10254–10261 (2011)', '  12', ' C', ' M', ' Summa, A', ' Lombardi, M', ' Lewis, W', ' F', ' DeGrado, Tertiary templates for the de-  sign of diiron proteins', ' Curr', ' Opin', ' Struct', ' Biol', ' 9, 500–508 (1999)', '  13', ' R', ' L', ' Koder et al', ', Nativelike structure in designed four alpha-helix bundles driven by  buried polar interactions', ' J', ' Am', ' Chem', ' Soc', ' 128, 14450–14451 (2006)', '  23', ' L', ' A', ' Lepak et al', ', “Handheld chem/biosensor using extreme conformational changes in designed binding proteins to enhance surface plasmon resonance” in Proceedings of SPIE: Advanced Environmental, Chemical, and Biological Sensing Technologies XIII, T', ' Vo-Dinh, R', ' A', ' Lieberman, G', ' G', ' Gauglitz, Eds', ' (The International Society for Optics and Photonics, Bellingham, WA, 2016), vol', ' 9862', '  24', ' L', ' Zhang, E', ' M', ' E', ' Andersen, A', ' Khajo, R', ' S', ' Magliozzo, R', ' L', ' Koder, Dynamic factors affecting gaseous ligand binding in an artificial oxygen transport protein', ' Bio- chemistry 52, 447–455 (2013)', '  14', ' G', ' V', ' Semisotnov et al', ', Study of the “molten globule” intermediate state in protein  25', ' J', ' Schleucher et al', ', A general enhancement scheme in heteronuclear multidimen-  folding by a hydrophobic fluorescent probe', ' Biopolymers 31, 119–128 (1991)', '  sional NMR employing pulsed field gradients', ' J', ' Biomol', ' NMR 4, 301–306 (1994)', '  Y G O L O I B  L A N O I T A T U P M O C  D N A  S C I S Y H P O I B     0 2 0 2    ,  5 2  I  I             y r a u r b e F n o S L A C D O R E P Y R A R B L E G E L L O C Y T C Y N U C     I  I          t  a   d e d a o n w o D  l  Schnatz et al', '  PNAS Latest Articles  | 7 of 7  \\x0c']\n"
     ]
    }
   ],
   "source": [
    "split_text = text.split(\".\")\n",
    "split_text = [s for s in split_text if len(s) != 0]\n",
    "split_text = [s.replace(\"\\n\", \" \") for s in split_text]\n",
    "print(split_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7100031337.0\n"
     ]
    }
   ],
   "source": [
    "def text2int(textnum, numwords={}):\n",
    "    ### https://stackoverflow.com/questions/493174/is-there-a-way-to-convert-number-words-to-integers\n",
    "    if not numwords:\n",
    "        units = [\n",
    "            \"zero\", \"one\", \"two\", \"three\", \"four\", \"five\", \"six\", \"seven\", \"eight\",\n",
    "            \"nine\", \"ten\", \"eleven\", \"twelve\", \"thirteen\", \"fourteen\", \"fifteen\",\n",
    "            \"sixteen\", \"seventeen\", \"eighteen\", \"nineteen\",\n",
    "            ]\n",
    "\n",
    "        tens = [\"\", \"\", \"twenty\", \"thirty\", \"forty\", \"fifty\", \"sixty\", \"seventy\", \"eighty\", \"ninety\"]\n",
    "\n",
    "        scales = [\"hundred\", \"thousand\", \"million\", \"billion\", \"trillion\"]\n",
    "\n",
    "        numwords[\"and\"] = (1, 0)\n",
    "        for idx, word in enumerate(units):    numwords[word] = (1, idx)\n",
    "        for idx, word in enumerate(tens):     numwords[word] = (1, idx * 10)\n",
    "        for idx, word in enumerate(scales):   numwords[word] = (10 ** (idx * 3 or 2), 0)\n",
    "\n",
    "    current = result = 0\n",
    "    for word in textnum.split():\n",
    "        if word not in numwords:\n",
    "            raise Exception(\"Illegal word: \" + word)\n",
    "\n",
    "        scale, increment = numwords[word]\n",
    "        current = current * scale + increment\n",
    "        if scale > 100:\n",
    "            result += current\n",
    "            current = 0\n",
    "\n",
    "    return float(result + current)\n",
    "\n",
    "print(text2int(\"seven billion one hundred million thirty one thousand three hundred thirty seven\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "units = [\n",
    "    \"zero\", \"one\", \"two\", \"three\", \"four\", \"five\", \"six\", \"seven\", \"eight\",\n",
    "    \"nine\", \"ten\", \"eleven\", \"twelve\", \"thirteen\", \"fourteen\", \"fifteen\",\n",
    "    \"sixteen\", \"seventeen\", \"eighteen\", \"nineteen\",\n",
    "    ]\n",
    "\n",
    "tens = [\"twenty\", \"thirty\", \"forty\", \"fifty\", \"sixty\", \"seventy\", \"eighty\", \"ninety\"]\n",
    "\n",
    "scales = [\"hundred\", \"thousand\", \"million\", \"billion\", \"trillion\"]\n",
    "\n",
    "number_words = units + tens + scales"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "original sentence: there are twenty three animals in the garden\n",
      "translated sentence: there are 23.0 animals in the garden\n",
      "****************************************************************************************************\n",
      "original sentence: i have a black cat\n",
      "original sentence: i don't want to live three million miles away\n",
      "translated sentence: i don't want to live 3000000.0 miles away\n",
      "****************************************************************************************************\n",
      "original sentence: talk to me for one minute\n",
      "translated sentence: talk to me for 1.0 minute\n",
      "****************************************************************************************************\n",
      "original sentence: i don't want any milk\n",
      "original sentence: you have one hundred and five problems\n",
      "translated sentence: you have 105.0 problems\n",
      "****************************************************************************************************\n",
      "original sentence: one hundred and two will not be the same as three thousand and one\n",
      "translated sentence: 102.0 will not be the same as 3001.0\n",
      "****************************************************************************************************\n",
      "original sentence: let's try just one hundred more times one million two hundred thousand and five more times but not more than two more times\n",
      "translated sentence: let's try just 100.0 more times 1200005.0 more times but not more than 2.0 more times\n",
      "****************************************************************************************************\n",
      "original sentence: the universe is 13,456 hundred years old\n",
      "translated sentence: the universe is 1345600.0 years old\n",
      "****************************************************************************************************\n",
      "original sentence: the earth is 4,500 years old and i am thirty three years old\n",
      "translated sentence: the earth is 4500.0 years old and i am 33.0 years old\n",
      "****************************************************************************************************\n",
      "original sentence: 3.14 is the beginning of pi\n",
      "translated sentence: 3.14 is the beginning of pi\n",
      "****************************************************************************************************\n",
      "original sentence: the earth is 13.5 billion years old\n",
      "translated sentence: the earth is 13500000000.0 years old\n",
      "****************************************************************************************************\n",
      "original sentence: i don't think this will work, 22 hundred million\n",
      "translated sentence: i don't think this will work, 2200000000.0\n",
      "****************************************************************************************************\n",
      "original sentence: there are 25 hundred people that are sick\n",
      "translated sentence: there are 2500.0 people that are sick\n",
      "****************************************************************************************************\n",
      "original sentence: i can't see past 1,525 thousand meters\n",
      "translated sentence: i can't see past 1525000.0 meters\n",
      "****************************************************************************************************\n"
     ]
    }
   ],
   "source": [
    "test_text_list = [\n",
    "                  \"there are twenty three animals in the garden\", \n",
    "                  \"I have a black cat\",\n",
    "                  \"I don't want to live three Million miles away\",\n",
    "                  \"talk to me for one minute\",\n",
    "                  \"I don't want any milk\",\n",
    "                  \"You have one hundred and five problems\",\n",
    "                  \"one hundred and two will not be the same as three thousand and one\",\n",
    "                  \"Let's try just one hundred more times one million two hundred thousand and five more times but\"\n",
    "                  \" not more than two more times\",\n",
    "                  \"The universe is 13,456 hundred years old\",\n",
    "                  \"the earth is 4,500 years old and I am thirty three years old\",\n",
    "                  \"3.14 is the beginning of pi\",\n",
    "                  \"the earth is 13.5 billion years old\",\n",
    "                  \"i don't think this will work, 22 hundred million\",\n",
    "                  \"there are 25 hundred people that are sick\",\n",
    "                  \"I can't see past 1,525 thousand meters\"]\n",
    "\n",
    "num_scale = {\"hundred\": 10**2,\n",
    "             \"thousand\": 10**3,\n",
    "             \"million\": 10**6,\n",
    "             \"billion\": 10**9,\n",
    "             \"trillion\": 10**12}\n",
    "\n",
    "translated_str = list()\n",
    "for s in test_text_list:\n",
    "    s = s.lower()\n",
    "    print(\"original sentence: {}\".format(s))\n",
    "    ### Does sentence contain integers or number words\n",
    "    if any(substring in s for substring in number_words) | bool(re.search(r'\\d', s)):\n",
    "    \n",
    "        word_split = s.split(\" \")\n",
    "        word_i = list()\n",
    "        used_num_words = list()\n",
    "        for idx, word in enumerate(word_split):\n",
    "            ### If numbers have commas, remove them\n",
    "            if re.search(r'\\d', word):\n",
    "                num_no_commas = float(word.replace(\",\", \"\"))\n",
    "                \n",
    "                j = 1\n",
    "                while word_split[idx + j] in number_words:\n",
    "                    num_no_commas = num_no_commas * num_scale[word_split[idx + j]]\n",
    "                    word = word + \" \" + word_split[idx + j]\n",
    "                    used_num_words.append(word_split[idx + j])\n",
    "                    j += 1\n",
    "                    \n",
    "                    if (idx + j) == len(word_split): break\n",
    "                    \n",
    "                s = s.replace(word, str(num_no_commas))\n",
    "\n",
    "            ### Get indices of number words not already translated\n",
    "            if word in used_num_words:\n",
    "                used_num_words.remove(word)\n",
    "            elif word in number_words:\n",
    "                word_i.append(idx)\n",
    "                \n",
    "        ### Check if there are any number words left to translate\n",
    "        if len(word_i) > 0:\n",
    "            all_word_nums_i = list()\n",
    "            ### there is a non number word in between number words\n",
    "            if (word_i[-1] - word_i[0]) != (len(word_i) - 1):\n",
    "\n",
    "                ### Split indices into lists of consecutive indices\n",
    "                consec_indices = list()\n",
    "                for k, g in groupby(enumerate(word_i), lambda i_x: i_x[0] - i_x[1]):\n",
    "                    consec_indices.append(list(map(itemgetter(1), g)))\n",
    "\n",
    "                ### combine lists of indices if they only skip one number\n",
    "                ### and the word corresponding to the missing index is \"and\"\n",
    "                for i, l in enumerate(consec_indices[:-1]):\n",
    "                    if (l[0] in [el for consec in all_word_nums_i for el in consec]):\n",
    "                        continue\n",
    "                    elif (consec_indices[i+1][0] - l[-1] == 2) & (word_split[l[-1] + 1] == \"and\"):\n",
    "                        all_word_nums_i.append(l + [l[-1] + 1] + consec_indices[i+1])\n",
    "                    else:\n",
    "                        all_word_nums_i.append(l)\n",
    "\n",
    "                if consec_indices[-1][0] not in [el for consec in all_word_nums_i for el in consec]:\n",
    "                    all_word_nums_i.append(consec_indices[-1])\n",
    "\n",
    "            else:\n",
    "                ### There is only one number with no non-number words within\n",
    "                all_word_nums_i = [word_i]\n",
    "\n",
    "            ### take each word list and translate it to an integer\n",
    "            for word_i in all_word_nums_i:\n",
    "                full_number = \" \".join([word_split[i] for i in word_i])\n",
    "                numeric_sub = text2int(full_number)\n",
    "                s = s.replace(full_number, str(numeric_sub))\n",
    "        \n",
    "        translated_str.append(s)\n",
    "        print(\"translated sentence: {}\".format(s))\n",
    "        print(\"*\"*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['there are 23.0 animals in the garden',\n",
       " 'i have a black cat',\n",
       " \"i don't want to live 3000000.0 miles away\",\n",
       " 'talk to me for 1.0 minute',\n",
       " \"i don't want any milk\",\n",
       " 'you have 105.0 problems',\n",
       " '102.0 will not be the same as 3001.0',\n",
       " \"let's try just 100.0 more times 1200005.0 more times but not more than 2.0 more times\",\n",
       " 'the universe is 1345600.0 years old',\n",
       " 'the earth is 4500.0 years old and i am 33.0 years old',\n",
       " '3.14 is the beginning of pi',\n",
       " 'the earth is 13500000000.0 years old',\n",
       " \"i don't think this will work, 2200000000.0\",\n",
       " 'there are 2500.0 people that are sick',\n",
       " \"i can't see past 1525000.0 meters\"]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translated_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
